### PR TITLE
Add user login history in reports

### PR DESCRIPTION
**Pull request by adudekyeti**
Add user login history in reports

### CODE REVIEW

File: app/TextParser/UserLoginHistory.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	 * Process.
+	 *
+	 * @return string
+	 */
+	public function process()
+	{
+		$existLoginHistory = false;
+		if (!empty($textParserParams = $this->textParser->getParam('textParserParams')) && isset($textParserParams['userId'])) {
+			$userId = $textParserParams['userId'];
+			$html = '<table><tr><td>' . \App\Language::translate('LBL_IP', 'Other:Reports') . '</td><td>' . \App\Language::translate('LBL_DATE') . '</td><td>' . \App\Language::translate('LBL_AGENT', 'Other:Reports') . '</td></tr>';
```
Comment (rskrzypczak) [inline app/TextParser/UserLoginHistory.php]: 'Other:Reports' ?

File: app/TextParser/UserLoginHistory.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+				->where([
+					'and',
+					['userid' => $userId],
+				]);
+			if (isset(\App\Condition::DATE_OPERATORS[$this->params[0]])) {
+				$dateRange = \DateTimeRange::getDateRangeByType($this->params[0]);
+				$historyLogin->andWhere(['between', 'login_time', $dateRange[0].' 00:00:00', $dateRange[1].' 23:59:59']);
+			}
+			$historyLogin = $historyLogin->all();
+			foreach ($historyLogin as $data) {
```
Comment (rskrzypczak) [inline app/TextParser/UserLoginHistory.php]: $dataReader
while()

File: modules/Users/models/Module.php [hunk 3]
```diff
@@ -102,6 +102,8 @@ public function saveLoginHistory($userName, $status = 'Signed in')
 				'logout_time' => null,
 				'status' => $status,
 				'browser' => $browser->name . ' ' . $browser->ver,
+				'userid' => \App\User::getUserIdByName($userName) ?? false,
```
Comment (rskrzypczak) [inline modules/Users/models/Module.php]:
```suggestion
				'userid' => \App\User::getUserIdByName($userName),
```
and chenge in getUserIdByName(): ?int

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: app/TextParser/UserLoginHistory.php (hunk 1 above)
Comment (adudekyeti) [reply app/TextParser/UserLoginHistory.php]: + agent to browser

File: app/TextParser/UserLoginHistory.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+				$existLoginHistory = true;
+				$html .= "<tr><td>{$data['user_ip']}</td><td>{$data['login_time']}</td><td>{$data['agent']}</td></tr>";
+			}
+			$html .= '</table>';
+		}
+		return $existLoginHistory ? $html : \App\Language::translate('LBL_NO_RECORDS', 'Other.Reports');
```
Comment (adudekyeti) [inline app/TextParser/UserLoginHistory.php]: ?

File: app/User.php [hunk 5]
```diff
@@ -415,15 +415,15 @@ public static function getActiveAdminId()
...
 	 * @return int
 	 */
-	public static function getUserIdByName($name)
+	public static function getUserIdByName($name): ?int
 	{
 		if (Cache::has(__METHOD__, $name)) {
 			return Cache::get(__METHOD__, $name);
 		}
 		$userId = (new Db\Query())->select(['id'])->from('vtiger_users')->where(['user_name' => $name])->limit(1)->scalar();
 		Cache::save(__METHOD__, $name, $userId, Cache::LONG);
```
Comment (rskrzypczak) [inline app/User.php]: ?
Cache::save(__METHOD__, $name, $userId, Cache::LONG);
$userId !== false ? $userId : null;

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

**Pull request merged**